src/sinks/greptimedb/service.rs

```diff
@@ -0,0 +1,188 @@
+
+        GreptimeDBRequest {
+            items,
+            finalizers,
+            metadata: request_metadata_builder.with_request_size(request_size),
+        }
+    }
+}
+
+impl Finalizable for GreptimeDBRequest {
+    fn take_finalizers(&mut self) -> EventFinalizers {
+        std::mem::take(&mut self.finalizers)
+    }
+}
+
+impl MetaDescriptive for GreptimeDBRequest {
+    fn get_metadata(&self) -> &RequestMetadata {
+        &self.metadata
+    }
+
+    fn metadata_mut(&mut self) -> &mut RequestMetadata {
+        &mut self.metadata
+    }
+}
+
+#[derive(Debug)]
+pub struct GreptimeDBBatchOutput {
+    pub item_count: u32,
+    pub metadata: RequestMetadata,
+}
+
+impl DriverResponse for GreptimeDBBatchOutput {
+    fn event_status(&self) -> EventStatus {
+        EventStatus::Delivered
+    }
+
+    fn events_sent(&self) -> &GroupedCountByteSize {
+        self.metadata.events_estimated_json_encoded_byte_size()
+    }
+
+    fn bytes_sent(&self) -> Option<usize> {
+        Some(self.metadata.request_wire_size())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct GreptimeDBService {
+    /// the client that connects to greptimedb
+    client: Arc<Database>,
+}
+
+impl GreptimeDBService {
+    pub fn try_new(config: &GreptimeDBConfig) -> crate::Result<Self> {
+        let grpc_client = if let Some(tls_config) = &config.tls {
+            let channel_config = ChannelConfig {
+                client_tls: Self::try_from_tls_config(tls_config)?,
+                ..Default::default()
+            };
+            Client::with_manager_and_urls(
+                ChannelManager::with_tls_config(channel_config).map_err(Box::new)?,
+                vec![&config.endpoint],
+            )
+        } else {
+            Client::with_urls(vec![&config.endpoint])
+        };
+
+        let mut client = Database::new_with_dbname(&config.dbname, grpc_client);
+
+        if let (Some(username), Some(password)) = (&config.username, &config.password) {
+            client.set_auth(AuthScheme::Basic(Basic {
+                username: username.to_owned(),
+                password: password.clone().into(),
+            }))
+        };
+
+        Ok(GreptimeDBService {
+            client: Arc::new(client),
+        })
+    }
+
+    fn try_from_tls_config(tls_config: &TlsConfig) -> crate::Result<Option<ClientTlsOption>> {
+        if let Some(ca_path) = tls_config.ca_file.as_ref() {
+            let cert_path = tls_config
+                .crt_file
+                .as_ref()
+                .ok_or(GreptimeDBConfigError::TlsMissingCert)?;
+            let key_path = tls_config
+                .key_file
+                .as_ref()
+                .ok_or(GreptimeDBConfigError::TlsMissingKey)?;
+
+            if tls_config.key_pass.is_some() {
+                warn!(
+                    message = "TLS key file with password is not supported by greptimedb client at the moment."
+                );
+            }
```

Comment:
If the `verify_certificate`, `verify_hostname` and `alpn_protocols` fields are ignored, can we add corresponding warning messages when they are set?

https://github.com/vectordotdev/vector/blob/b91976da24f75150fb800fa11deea620e4b5abbb/lib/vector-core/src/tls/settings.rs#L87